File: src/AetherNet.Core/Identity/EridAnnounceService.cs
// SPDX-License-Identifier: MIT

using AetherNet.Constants;
using AetherNet.Protocol;
using AetherNet.Routing;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace AetherNet.Identity;

/// <summary>Event args: an inbound ERID announcement (still Signal-encrypted) plus the peer that sent it.</summary>
public sealed class EridAnnounceReceived : EventArgs
{
    /// <summary>The packet body — a Signal <c>EncryptedPayload</c> whose plaintext is an <see cref="EridAnnouncementCodec"/> frame.</summary>
    public byte[] EncryptedAnnouncement { get; init; } = Array.Empty<byte>();

    /// <summary>UHID of the peer that sent the announcement.</summary>
    public string FromUhid { get; init; } = string.Empty;
}

/// <summary>
/// Binds <see cref="PacketType.EridAnnounce"/> (56) to the mesh: a node shares its rotating-address
/// routing key with an established peer by sending the (already Signal-encrypted) announcement directly.
/// Transport only — the plaintext framing (<see cref="EridAnnouncementCodec"/>) and the encryption
/// (ISignalProtocolService) are done by the host/EridExchangeService; this service just carries the
/// opaque encrypted blob as a directed packet and surfaces inbound ones via <see cref="AnnounceReceived"/>.
/// </summary>
public interface IEridAnnounceService
{
    /// <summary>Raised when an ERID announcement arrives from a peer (payload still encrypted).</summary>
    event EventHandler<EridAnnounceReceived>? AnnounceReceived;

    /// <summary>Send an encrypted ERID announcement directly to <paramref name="peerUhid"/>. Returns delivery success.</summary>
    Task<bool> SendAnnounceAsync(string peerUhid, byte[] encryptedAnnouncement, CancellationToken cancellationToken = default);

    /// <summary>Process an inbound <see cref="PacketType.EridAnnounce"/>: raise <see cref="AnnounceReceived"/>. False on wrong type or empty body.</summary>
    Task<bool> HandleAsync(MeshPacket packet, CancellationToken cancellationToken = default);
}

/// <inheritdoc />
public sealed class EridAnnounceService : IEridAnnounceService
{
    private readonly IMeshSender _sender;
    private readonly ILogger<EridAnnounceService> _logger;

    public event EventHandler<EridAnnounceReceived>? AnnounceReceived;

    public EridAnnounceService(IMeshSender sender, ILogger<EridAnnounceService>? logger = null)
    {
        _sender = sender ?? throw new ArgumentNullException(nameof(sender));
        _logger = logger ?? NullLogger<EridAnnounceService>.Instance;
    }

    /// <inheritdoc />
    public async Task<bool> SendAnnounceAsync(string peerUhid, byte[] encryptedAnnouncement, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(peerUhid);
        ArgumentNullException.ThrowIfNull(encryptedAnnouncement);
        if (encryptedAnnouncement.Length == 0)
            throw new ArgumentException("encryptedAnnouncement cannot be empty.", nameof(encryptedAnnouncement));

        var packet = new MeshPacket
        {
            Type = PacketType.EridAnnounce,
            SourceUhid = _sender.LocalUhid,
            DestinationUhid = peerUhid,
            Ttl = ProtocolConstants.DefaultTtl,
            Payload = encryptedAnnouncement,
        };
        var delivered = await _sender.SendAsync(packet, peerUhid, cancellationToken).ConfigureAwait(false);
        _logger.LogDebug("ERID announce → {Peer} ({Bytes} B) delivered={Delivered}", peerUhid, encryptedAnnouncement.Length, delivered);
        return delivered;
    }

    /// <inheritdoc />
    public Task<bool> HandleAsync(MeshPacket packet, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(packet);
        if (packet.Type != PacketType.EridAnnounce)
            return Task.FromResult(false);
        if (packet.Payload is null || packet.Payload.Length == 0)
            return Task.FromResult(false);

        AnnounceReceived?.Invoke(this, new EridAnnounceReceived
        {
            EncryptedAnnouncement = packet.Payload,
            FromUhid = packet.SourceUhid,
        });
        return Task.FromResult(true);
    }
}

File: typescript/src/eridannounce/EridAnnounceService.ts
/**
 * ERID-announce mesh binding (PacketType.EridAnnounce = 56).
 *
 * A node shares its rotating-address routing key with an established peer by sending the
 * (already Signal-encrypted) ERID announcement directly. Transport only — the plaintext framing
 * (EridAnnouncementCodec) and the encryption (SignalProtocol) are done by the host/EridExchange
 * layer; this service just carries the opaque encrypted blob as a directed packet and surfaces
 * inbound ones via onAnnounceReceived.
 *
 * Mirrors the C# EridAnnounceService.
 *
 * SPDX-License-Identifier: MIT
 */

import { DEFAULT_TTL } from "../constants.js";
import { MeshPacket } from "../protocol/MeshPacket.js";
import { PacketType } from "../protocol/PacketType.js";
import { IMeshSender } from "../routing/IMeshSender.js";

/**
 * Binds PacketType.EridAnnounce (56) to the mesh. Send an already-encrypted ERID announcement
 * directly to a peer, and surface inbound announcements (payload still encrypted) via
 * onAnnounceReceived.
 */
export class EridAnnounceService {
  /**
   * Raised when an ERID announcement arrives from a peer (payload still Signal-encrypted). The
   * host decrypts it and feeds the plaintext to EridAnnouncementCodec.tryDecode.
   *
   * @param encryptedAnnouncement the opaque encrypted announcement bytes (the packet body).
   * @param fromUhid UHID of the peer that sent the announcement.
   */
  onAnnounceReceived?: (encryptedAnnouncement: Uint8Array, fromUhid: string) => void;

  constructor(private readonly sender: IMeshSender) {}

  /**
   * Send an already-encrypted ERID announcement directly to `peerUhid`: build a directed
   * EridAnnounce packet carrying the opaque blob. Returns delivery success. Throws for an empty
   * peer uhid or an empty announcement.
   */
  async sendAnnounce(peerUhid: string, encryptedAnnouncement: Uint8Array): Promise<boolean> {
    if (!peerUhid) throw new Error("peerUhid must not be empty");
    if (encryptedAnnouncement.length === 0) {
      throw new Error("encryptedAnnouncement cannot be empty");
    }

    const packet = new MeshPacket();
    packet.type = PacketType.EridAnnounce;
    packet.sourceUhid = this.sender.localUhid;
    packet.destinationUhid = peerUhid;
    packet.ttl = DEFAULT_TTL;
    packet.payload = encryptedAnnouncement;

    return this.sender.send(packet, peerUhid);
  }

  /**
   * Process an incoming PacketType.EridAnnounce packet: surface it via onAnnounceReceived. Returns
   * false for the wrong packet type or an empty body.
   */
  async handle(packet: MeshPacket): Promise<boolean> {
    if (packet.type !== PacketType.EridAnnounce) return false;
    if (packet.payload.length === 0) return false;

    this.onAnnounceReceived?.(packet.payload, packet.sourceUhid);
    return true;
  }
}
